ehr_billing/src/org/labkey/ehr_billing/notification/BillingNotification.java

```diff
@@ -410,13 +407,18 @@ protected void createChargeSummaryReport(final StringBuilder msg, Date lastInvoi
         msg.append("The table below summarizes projected charges since the last invoice date of " + getDateFormat(c).format(lastInvoiceEnd));
 
         msg.append("<table border=1 style='border-collapse: collapse;'><tr style='font-weight: bold;'><td align='center'>Category</td><td align='center'># Items</td><td align='center'>Amount</td>");
-        for (String category : totalsByCategory.keySet())
+        for (String category : categoryInfoMap.keySet())
         {
-            Map<String, Double> totalsMap = totalsByCategory.get(category);
+            ChargeCategoryInfo categoryInfo = categoryInfoMap.get(category);
             Container container = containerMap.get(category);
-
-            String url = createURL(container, centerSpecificBillingSchema, categoryToQuery.get(category), null) + "&query.param.StartDate=" + getDateFormat(c).format(start.getTime()) + "&query.param.EndDate=" + getDateFormat(c).format(endDate.getTime());
-            msg.append("<tr><td><a href='" + url + "'>" + category + "</a></td><td align='right'>" + _numItemsFormat.format(totalsMap.get("total")) + "</td><td align='right'>" + _dollarFormat.format(totalsMap.get("totalCost")) + "</td></tr>");
+            String url;
+            if (categoryToQuery.containsKey(category))
+                url = createURL(container, centerSpecificBillingSchema, categoryToQuery.get(category), null) + "&query.param.StartDate=" + getDateFormat(c).format(start.getTime()) + "&query.param.EndDate=" + getDateFormat(c).format(endDate.getTime());
+            else
+                url = categoryInfo.getCategoryURL();
+
+            if (null != url)
+                msg.append("<tr><td><a href='" + url + "'>" + category + "</a></td><td align='right'>" + _numItemsFormat.format(categoryInfo.getTotalQuantity()) + "</td><td align='right'>" + _dollarFormat.format(categoryInfo.getTotalCost()) + "</td></tr>");
```

Comment:
I missed this before. `category` should be HTML-encoded.

Reply:
`url` and even formatted values should be HTML encoded as well. It's easy to set format patterns that produce HTML-unsafe results.

Reply:
Ah yup! thank you, both!